google/cloud/internal/backoff_policy.h

```diff
@@ -0,0 +1,151 @@
+ * flexibility, and also provide sensible default implementations.
+ *
+ * The client library receives an object of this type, and clones a new instance
+ * for each operation. That is, the application provides the library with a
+ * [Prototype](https://en.wikipedia.org/wiki/Prototype_pattern) of the policy
+ * that will be applied to each operation.
+ *
+ * [Truncated Exponential
+ * Backoff](https://cloud.google.com/storage/docs/exponential-backoff) in the
+ * Google Cloud Storage documentation.
+ *
+ */
+class BackoffPolicy {
+ public:
+  virtual ~BackoffPolicy() = default;
+
+  /**
+   * Return a new copy of this object.
+   *
+   * Typically implemented as
+   * @code
+   *   return std::unique_ptr<BackoffPolicy>(new Foo(*this));
+   * @endcode
+   */
+  virtual std::unique_ptr<BackoffPolicy> clone() const = 0;
+
+  /**
+   * Return the delay after an RPC operation has completed.
+   *
+   * @return how long to wait before trying the operation again.
+   */
+  virtual std::chrono::milliseconds OnCompletion() = 0;
+};
+
+/**
+ * Implements a truncated exponential backoff with randomization policy.
+ *
+ * This policy implements the truncated exponential backoff policy for
+ * retrying operations. After a request fails, and subject to a separate
+ * retry policy, the client library will wait for an initial delay before
+ * trying again. If the second attempt fails the delay time is increased,
+ * using an scaling factor. The delay time growth stops at a maximum delay
+ * wait time. The policy also randomizes the delay each time, to avoid
+ * [thundering herd
+ * problem](https://en.wikipedia.org/wiki/Thundering_herd_problem).
+ */
+class ExponentialBackoffPolicy : public BackoffPolicy {
+ public:
+  /**
+   * Constructor for an exponential backoff policy.
+   *
+   * Define the initial delay, maximum delay, and scaling factor for an instance
+   * of the policy. While the constructor accepts `std::chrono::duration`
+   * objects at any resolution, the data is kept internally in microseconds.
+   * Sub-microsecond delays seem unnecessarily precise for this application.
+   *
+   * @code
+   * using namespace std::chrono_literals; // C++14
+   * auto r1 = ExponentialBackoffPolicy<S,T>(10ms, 500ms);
+   * auto r2 = ExponentialBackoffPolicy<S,T>(10min, 10min + 2s);
+   * @endcode
+   *
+   * @param initial_delay how long to wait after the first (unsuccessful)
+   *     operation.
+   * @param maximum_delay the maximum value for the delay between operations.
+   * @param scaling how fast does the delay increase between iterations.
+   *
+   * @tparam Rep1 a placeholder to match the Rep tparam for @p initial_delay's
+   *     type, the semantics of this template parameter are documented in
+   *     `std::chrono::duration<>` (in brief, the underlying arithmetic type
+   *     used to store the number of ticks), for our purposes it is simply a
+   *     formal parameter.
+   * @tparam Rep2 similar formal parameter for the type of @p maximum_delay.
```

Comment:
Nit: These are documented in a different order than they're defined below.

Reply:
Fixed.